### ISSUE TITLE
`Process_Builder` hangs if program output is too large

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#8777]
I have found out that our Enso `Process_Builder` hangs if the program that it runs returns too large output. Note that the threshold when this starts happening is not very large - it seems to be somewhere around 4-10kB (possibly OS-specific).

Repro:
```
from Standard.Base import all

prog = """
    const n = parseInt(process.argv[2]);
    for (var i = 0; i < n; i++) {
        console.log("A");
    }

run_proc n =
    IO.println "Running with arg="+n.to_text
    result = Process.new_builder "node" ["printer.js", n.to_text] . create
    result_n = result.stdout.length
    IO.println "   ==> has returned "+result_n.to_text

main =
    prog.write (File.new "printer.js")
    [1, 20, 100, 200, 500, 1000, 2000, 4000, 9000, 30000].each run_proc
```

for me yields the following results:
```
Running with arg=1
   ==> has returned 2
Running with arg=20
   ==> has returned 40
Running with arg=100
   ==> has returned 200
Running with arg=200
   ==> has returned 400
Running with arg=500
   ==> has returned 1000
Running with arg=1000
   ==> has returned 2000
Running with arg=2000
   ==> has returned 4000
Running with arg=4000
<HERE the program hangs forever>
```

**Comment by radeusgd** [comment]
I think I've already encountered a similar behaviour in the past.

This happens only if we don't inherit IO, but redirect the stdout of the child process to a Pipe that is connected to a `ByteArrayOutputStream`.

The problem is a kind of a 'deadlock':
https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/system/System.java#L112-L126

Our `System.create_process` builtin does the following sequence of operations:
1. Write ALL provided input to the child process's stdin (if specified).
2. Wait until the child process exits (`waitFor`).
3. Read all of stdout that the process wrote.
4. Read all of stderr that the process wrote.

However, this means that the other end of the stdout Pipe is only read after the process exits. If the process writes a lot of data to the Pipe, it gets 'saturated' and further writes are blocked by the OS, waiting for the other side of the Pipe to 'unclog' and read something from it. But this never happens, because Enso waits until the process exits before it will start reading. But the other process will not exit before it can finish the write operation, that needs to 'free buffer space' in the Pipe. So we get a deadlock.

**Comment by radeusgd** [comment]
To fix this, we need to be reading the stdout and stderr of the child process concurrently, a bit similarly to how I did it in one of our old tests:
https://github.com/enso-org/enso/blob/develop/lib/scala/testkit/src/main/scala/org/enso/testkit/process/WrappedProcess.scala#L23-L52

This way, we can buffer the read data on Java heap instead of within the Pipe that usually has a small buffer, and thus the child process will be able to continue writing even large chunks of output.

We may want to consider limiting the size of the buffered output to avoid OOM, but in case we exceed it, we should not block the child process but close the stream to indicate to it that no further writes will succeed. In such case, we should probably throw an exception.

**Comment by radeusgd** [comment]
I suspect this may be causing CI failures for me:
https://github.com/enso-org/enso/actions/runs/7542291021/job/20565495886?pr=8761#step:10:13702

```
 INFO ide_ci::program::command: cmdℹ️ Found 2 cases to execute (ETA 50s)
 INFO ide_ci::program::command: cmdℹ️ Benchmarking 'Startup.empty_startup' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={3 iterations, 5 seconds each}]
 INFO ide_ci::program::command: cmdℹ️ [DRY-RUN] Benchmark 'Startup.empty_startup' finished in 3945.615 ms
 INFO ide_ci::program::command: cmdℹ️ Benchmarking 'Startup.hello_world_startup' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={3 iterations, 5 seconds each}]
```
Here the process seems to hang, in another run it exceeded the 6h time limit.
In this example I hit cancel after waiting for a while, causing the following to be printed:
```
error: process didn't exit successfully: `target/rust/buildscript\enso-build-cli.exe backend ci-check` (exit code: 0xc000013a, STATUS_CONTROL_C_EXIT)
Error: The operation was canceled.
```

Although I'm not sure why they weren't occurring before.

<blockquote><img src="https://opengraph.githubassets.com/63841532db4796f90244ed233d2a9eef4ceb21a72ffe48a0144b10fbc7c86caf/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Allow secrets in `Header.authorization_*` · enso-org/enso@3323131</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by Akirathan** [comment]
Duplicate of #9549
